src/libraries/metadatacheckstatehandler.cpp

```diff
@@ -0,0 +1,188 @@
+** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+** GNU General Public License for more details.
+**
+** You should have received a copy of the GNU General Public License
+** along with this program.  If not, see <http://www.gnu.org/licenses/>.
+**
+****************************************************************************/
+
+#include "metadatacheckstatehandler.h"
+
+#include "metadatamodel.h"
+
+using namespace Asn1Acn::Internal::Libraries;
+
+MetadataCheckStateHandler::MetadataCheckStateHandler(const MetadataModel *model)
+    : m_model(model)
+{
+}
+
+const MetadataCheckStateHandler::States &MetadataCheckStateHandler::changedIndexes() const
+{
+    return m_changedIndexes;
+}
+
+const MetadataCheckStateHandler::Conflict &MetadataCheckStateHandler::conflict() const
+{
+    return m_conflict;
+}
+
+bool MetadataCheckStateHandler::changeCheckStates(const QModelIndex &index, const Qt::CheckState state)
+{
+    if (state == Qt::Unchecked)
+        return uncheck(index);
+
+    if (state == Qt::Checked)
+        return check(index, state);
+
+    return true;
+}
+
+bool MetadataCheckStateHandler::uncheck(const QModelIndex &index)
+{
+    const auto state = Qt::Unchecked;
+
+    m_changedIndexes.insert(index, state);
+
+    updateChildren(index, state);
+    updateParent(index, state);
+
+    return true;
+}
+
+bool MetadataCheckStateHandler::check(const QModelIndex &index, const Qt::CheckState state)
+{
+    if (m_changedIndexes.contains(index) && m_changedIndexes.value(index) == state)
+        return true;
+
+    m_changedIndexes.insert(index, state);
+
+    if (!updateParent(index, state) || !updateChildren(index, state) || !handleRelatives(index))
+        return false;
+
+    if (!m_relatives.isEmpty())
+        return check(m_relatives.takeFirst(), state);
+
+    return true;
+}
+
+bool MetadataCheckStateHandler::updateChildren(const QModelIndex &index, const Qt::CheckState state)
+{
+    for (auto i = 0; i < m_model->rowCount(index); ++i) {
+        for (auto j = 0; j < m_model->columnCount(index); ++j) {
+            auto child = index.child(i, j);
+
+            m_changedIndexes.insert(child, state);
+
+            if (state != Qt::Unchecked && !handleRelatives(child))
+                return false;
+
+            if (!updateChildren(child, state))
+                return false;
+        }
+    }
+
+    return true;
+}
+
+bool MetadataCheckStateHandler::updateParent(const QModelIndex &index, const Qt::CheckState childState)
+{
+    auto state = parentState(index, childState);
+
+    for (auto parent = index.parent(); parent.isValid(); parent = parent.parent()) {
+        m_changedIndexes.insert(parent, state);
+        if (state != Qt::Unchecked && !handleRelatives(parent))
+            return false;
+
+        state = parentState(parent, state);
+    }
+
+    return true;
+}
+
+bool MetadataCheckStateHandler::handleRelatives(const QModelIndex &index)
+{
+    enqueueRequired(index);
+    return noConflicts(index);
```

Comment:
`hasNoConflicts`? I like when there's a verb in function name :) (when it's not an obvious getter)

Reply:
Approved, renamed.